test/python/test_certificate_authority_api.py

```diff
@@ -0,0 +1,297 @@
+from __future__ import absolute_import
+
+import pathlib
+import unittest
+
+import openapi_client
+
+import api_config
+
+CA_POLICY = pathlib.Path('test/config/cert_auth_policy.yml')
+CERT_CHAIN_PATH = pathlib.Path('config/https/ca.crt')
+UNENCRYPTED_KEY_PATH = pathlib.Path('config/https/intermediate.key')
+ENCRYPTED_KEY_PATH = pathlib.Path('config/https/intermediate_encrypted.key')
+KEY_PASSWORD_PATH = pathlib.Path('config/https/intermediate_key_password.txt')
+CSR_PATH = pathlib.Path('test-python.csr')
+
+def read_file(path):
+    """Returns the entirety of a file's contents"""
+    with open(path, 'r') as content:
+        return content.read()
+
+class TestCertificateAuthorityApi(api_config.ConfiguredTest):
+    """CertificateAuthorityApi unit test stubs"""
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+
+        cls.CA_SERVICE_ID = 'signing-service'
+
+        cls.TEST_CLIENT_ID    = f'{cls.account}:host:{cls.CA_SERVICE_ID}/test-client'
+        cls.CN_TEST_CLIENT_ID = f'{cls.account}:host:{cls.CA_SERVICE_ID}/cn-test-client'
+        cls.NO_SIGN_CLIENT_ID = f'{cls.account}:host:{cls.CA_SERVICE_ID}/no-sign-client'
+
+        # read generated CSR
+        cls.csr = read_file(CSR_PATH)
+
+    def setUp(self):
+        # load CA policy as root
+        # creates a CA webservice,
+        # a group with sign privilege on the service,
+        # and host clients with varying privilege
+        ca_policy = read_file(CA_POLICY)
+        policy_api = openapi_client.PoliciesApi(self.client)
+        new_role_details = policy_api.load_policy(
+            self.account,
+            'root',
+            ca_policy
+        )
+
+        created_roles = new_role_details['created_roles']
+        self.test_client_api_key = created_roles[self.TEST_CLIENT_ID]['api_key']
+        self.cn_test_client_api_key = created_roles[self.CN_TEST_CLIENT_ID]['api_key']
+        self.no_sign_client_api_key = created_roles[self.NO_SIGN_CLIENT_ID]['api_key']
+
+        # set up Certificate Authority API client
+        client = api_config.get_api_client(
+            'host/signing-service/test-client',
+            self.test_client_api_key
+        )
+        self.api = openapi_client.CertificateAuthorityApi(client)
+        self.bad_auth_api = openapi_client.CertificateAuthorityApi(self.bad_auth_client)
+
+        # assign intermediate CA private key and CA chain to Conjur variables
+        ca_chain = read_file(CERT_CHAIN_PATH)
+        private_key = read_file(UNENCRYPTED_KEY_PATH)
+
+        self.secrets_api = openapi_client.SecretsApi(self.client)
+        self.configure_ca(ca_chain, private_key, None)
+
+    def configure_ca(self, ca_chain=None, private_key=None, key_password=None):
```

Comment:
This seems like three methods stacked on top of each other underneath a trench coat. I think they might be better served as individuals. It would reduce the need for the three kwargs and the if statements. If you were to split it the `secrets_api` variable would make sense as a class field